### ISSUE TITLE
Remove or add redundant phpdoc

### ISSUE THREAD

**Issue by theofidry** [infection/infection#861]
Basically we need to choose between:

```php
class X {

  private $foo;

  public function __construct(string $foo) {
    $this->foo = $foo;
  }

  public function getFoo(): string {
    return $this->foo;
  }
}
```

or:

```php
class X {

  /**
   * @var string
   */
  private $foo;

  public function __construct(string $foo) {
    $this->foo = $foo;
  }

  public function getFoo(): string {
    return $this->foo;
  }
}
```

This does not apply to properties which are not set in the constructor.

**Comment by maks-rafalko** [comment]
I would say we need to force it by fixer, otherwise it will be inconsistent anyway (human factor). I didn't find any rule for PHP-CS-Fixer to remove (or add) property types. Are you aware of one?

I found [something similar in Rector](https://www.tomasvotruba.cz/blog/2019/01/03/how-to-complete-type-declarations-without-docblocks-with-rector/#properties-without-var), but didn't use it personally.

Also, could be useful for us in the future: [`TypedPropertyRector`](https://github.com/rectorphp/rector/blob/master/docs/AllRectorsOverview.md#typedpropertyrector)

```diff
final class SomeClass
 {
-    /**
-     * @var int
-     */
-    private count;
+    private int count;
 }
```

Another one: [`CompleteVarDocTypePropertyRector`](https://github.com/rectorphp/rector/blob/master/docs/AllRectorsOverview.md#completevardoctypepropertyrector)

```diff
final class SomeClass
 {
+    /**
+     * @var EventDispatcher
+     */
     private $eventDispatcher;

     public function __construct(EventDispatcher $eventDispatcher)
     {
         $this->eventDispatcher = $eventDispatcher;
     }
 }
```

And one more: [`PropertyTypeDeclarationRector`](https://github.com/rectorphp/rector/blob/master/docs/AllRectorsOverview.md#propertytypedeclarationrector)

---

That being said, we have some automatic tools that can convert 

* PHPDoc to PHP 7.4 property type declarations
* Constructor parameters to property PHPDoc type hint (constructor parameter type inference)

---

Another thing is with `array` properties. From my point of view, even if we decide to remove PHPDoc property typehints, `array` properties *must* have a concrete type (e.g. `int[]`). And [there is a PHPStan rule to control it](https://github.com/thecodingmachine/phpstan-strict-rules/blob/e3d746a61d38993ca2bc2e2fcda7012150de120c/doc/typehinting_rules.md#vague-array-type-hint-detection)

TL;DR (my opinion)

* this should be controlled by automatic tool (add or remove)
* we have to be able to convert types to native property type declarations in the future
* we must add type hints for `array` properties

**Comment by sanmai** [comment]
If we can't remove it, the very least we can check for its presence with a simple `grep` and then a failing build.

**Comment by theofidry** [comment]
I think we should first decide on what direction we are taking:

1. Removing those docs
2. Adding/Keeping those docs

I'm for 1. but I think we should decide as a team

**Comment by theofidry** [comment]
ping @infection/core

**Comment by BackEndTea** [comment]
PHPStan can infer the types from private properties, but not from protected or public ones. So whichever way we go, we do need to keep them on those properties.

**Comment by theofidry** [comment]
Good to know 👍

**Comment by maks-rafalko** [comment]
> PHPStan can infer the types from private properties, but not from protected or public ones. So whichever way we go, we do need to keep them on those properties.

Then, this should be definitely check by some automated tools if we decide to remove PHPDoc block from private properties. Otherwise we will 100% end up with phpdoc being removed from (or not being added to) protected/public properties
